numpy/core/shape_base.py

```diff
@@ -455,13 +455,31 @@ def atleast_nd(a, ndim):
     def block_recursion(self, arrays, depth=0):
         if depth < max_depth:
             arrs = [self(arr, depth+1) for arr in arrays]
-            return _nx.concatenate(arrs, axis=-(max_depth-depth))
+            if len(arrs) == 1:
+                # concatenate copies the array.
+                # Avoid calling it if we don't need to
+                return arrs[0]
+            else:
+                return _nx.concatenate(arrs, axis=-(max_depth-depth))
         else:
             # We've 'bottomed out' - arrays is either a scalar or an array
             # type(arrays) is not list
             return atleast_nd(arrays, result_ndim)
-
-    return block_recursion(arrays)
+    result = block_recursion(arrays)
+
+    # Block should return a copy of that array and not a view.
+    # Catch an edge case where the user provies a single array to block.
+    # The user may have passed the array deep inside a nested list.
+    first_array = arrays
+    for _ in range(max_depth):
+        first_array = first_array[0]
+    # The user may have provided an array that is already of the correct shape
+    # Therefore, the `first_array is result`
+    # Otherwise, the `first_array` may have been upcast by `atleast_nd`
+    # In which case, we need to check `result.base`
+    if (first_array is result or first_array is result.base):
```

Comment:
Inspecting `.base` like this is rarely a good idea.

Reply:
Is there a good way to check if two arrays point to the same data?

Reply:
`np.may_share_memory`, but it's slow. The best way in this case is just to keep track of it yourself.

Reply:
got it.